---
title: "08_MetWAS"
author: "Jagadeesh Puvvula"
date: "2025-01-15"
output: pdf_document
---
#work from line 117
#air pollution and meta data
```{r}
#air pollution data - standardized to mean 0 and SD=1
load("~/Documents/air_polln_rna_seq/crib_data/data_clean_dec11_2024/exposure_data.rda")
rm(list = setdiff(ls(), "ap_exposures"))
pm1<- ap_exposures |>
  select(subject_id, 
         matches("^pm_([1-9]|[1-9][0-9]|1[0-3][0-9]|140|141)$"),
         matches("^bc_([1-9]|[1-9][0-9]|1[0-3][0-9]|140|141)$")) |>
  mutate(across(
    .cols = -subject_id, 
    .fns = ~ scale(.x, center = TRUE, scale = TRUE)
  )) |>
  clean_names()


#meta data
load("~/Documents/air_polln_rna_seq/crib_data/data_clean_dec11_2024/df_analysis.rda") 
df_analysis<- df_analysis |> select(-c(5,13,14)) |>
  filter(!(subject_id == 1092 & duplicated(subject_id)),
         subject_id != 389) |> #missing RNAseq data
  left_join(pm1, by = "subject_id") 

rm(list = setdiff(ls(), "df_analysis"))
```


#clean metabolome data
```{r}
metabo_linkage <- read_csv("~/Documents/air_polln_rna_seq/crib_data/metabolome_paired/metabolome_linkage.csv") |>
  clean_names() |>
  select(c(1,23)) |>
  mutate(sample_description = sub("PCFSBDEB[1-4]$", "", sample_description))

#metabolite annotations
df_annotations <- read_csv("~/Documents/air_polln_rna_seq/crib_data/metabolome_paired/metab_annot.csv") |>
  clean_names()

#original data - batch normalized and imputed data contain 866 metabolites
#10 metabolites dropped as all of them have same concentrations
placenta_metabo <- read_csv("~/Documents/air_polln_rna_seq/crib_data/metabolome_paired/metabolome_conc.csv") |>
  clean_names() |>
  left_join(metabo_linkage, by="parent_sample_name") |>
  mutate(sample_description = as.numeric(sample_description)) |> 
  rename(subject_id = sample_description) |> 
  select(subject_id, everything()) |>
  select(-c(2)) |>
  mutate(across(2:last_col(), ~ log10(. + 0.00001))) |>
  mutate(across(2:last_col(), ~ (. - mean(.)) / sd(.))) |>
  select(-where(~any(is.na(.)))) |>
  filter(subject_id %in% df_analysis$subject_id)

transposed_df <- placenta_metabo |>
  t() |>
  as_tibble(rownames = "Variable") |>
  row_to_names(row_number = 1) |>
  rename(chem_id = 1) |>
  mutate(chem_id = sub("^x", "", chem_id))
```

#save metabolome and exposure dat for analysis
```{r}
save(df_analysis, transposed_df, df_annotations,
     file = "~/Documents/air_polln_rna_seq/crib_data/metabolome_paired/df_metabolome_analysis.rda")
```

#MetWAS analysis
```{r}
load("~/Documents/air_polln_rna_seq/crib_data/metabolome_paired/df_metabolome_analysis.rda")

x<- df_analysis$subject_id

#sort metabolome data
transposed_df <- transposed_df |> 
  select(1, all_of(intersect(as.character(x), colnames(transposed_df)[-1])))
```

#check the sequence and participants with pm data have metabolome data
```{r}
x <- as.character(df_analysis$subject_id)
cat("All values present:", all(x %in% colnames(transposed_df)[-1]), "\n", "Sequence matches:", identical(x, colnames(transposed_df)[-1]), "\n")
```

#MetWAS
```{r}
appliedepi::mwas_lm_loop(feature_table = transposed_df, 
                         exp_cov_data = df_analysis,
                         exposures = names(df_analysis)[13:294],
                         covar = c("race", "maternal_age", "bmi", "education",
                                   "parity", "tobacco_use_dur_preg"), 
                         output_folder = "~/Documents/air_polln_rna_seq/results/lm_MetWAS",
                         mwas_file_name="placenta_mwas.csv",
                         fdr_cutoff= 0.2)
```

#load mwas results
```{r}
#beta_dir variable is based on 20% FDR cutoff
res_df<- read_csv("~/Documents/air_polln_rna_seq/results/lm_MetWAS/placenta_mwas.csv") |>
  clean_names() |>
  left_join(
    df_annotations |>
      select(1, 11, 6, 5), 
    by = "chem_id"
  )

write_csv(res_df, "~/Documents/air_polln_rna_seq/results/lm_MetWAS/placenta_mwas_anno.csv")
```
############################################
############################################
############################################
############################################
#summary of results
```{r}
res_df<- read_csv("~/Documents/air_polln_rna_seq/results/lm_MetWAS/placenta_mwas_anno.csv")

summary_df <- res_df |>
  #filter(beta_dir %in% c("negative-significant", "positive-significant")) |>
  filter(fdr<0.05) |>
  group_by(variable, super_pathway, sub_pathway, beta_dir) |>
  summarise(count = n(), .groups = "drop") |>
  filter(count>1) |>
  mutate(exposure = case_when(
    str_starts(variable, "bc") ~ "Black carbon",
    str_starts(variable, "pm") ~ "PM2.5",
    TRUE ~ NA_character_  # Default case if neither condition is met
  ),
  exposure_window = gsub("^(pm_|bc_)", "", variable),
  exposure_window = factor(exposure_window, 
                             levels = sort(unique(as.numeric(gsub("^(pm_|bc_)", "", variable)))))
  ) |>
  mutate(super_pathway = recode(super_pathway, 
                                "Partially Characterized Molecules" = "Other",
                                "Energy" = "Other"),
         super_pathway = fct_relevel(super_pathway, 
                                     "Lipid", 
                                     "Amino Acid", 
                                     "Carbohydrate", 
                                     "Cofactors and Vitamins", 
                                     "Peptide", 
                                     "Nucleotide", 
                                     "Xenobiotics",
                                     "Other"))
```

#visualizing results - Summarized heat map
```{r}
ggplot(summary_df, aes(x = exposure_window, y = sub_pathway, fill = count)) +
  facet_grid(super_pathway~exposure, scales="free", space="free",
             switch = "both",
             labeller = label_wrap_gen(width = 15)) +
  geom_tile(color = "white") + 
  scale_fill_viridis(name = "Number of metabolites") +
  labs(x = "Gestational period (days)", y = "Metabolic pathways") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, color = "black"),
        axis.text.y = element_text(size=11, color = "black"),
        legend.position = "bottom",
        legend.text = element_text(angle = 45, hjust = 1, color = "black"),
        legend.title = element_text(vjust = 1),
        legend.key.height = unit(0.3, "cm"),
        panel.grid.major.x = element_blank(), 
        panel.grid.minor.x = element_blank(),
        strip.text.y = element_text(size=9, angle = 45, hjust = 0, colour = "black"),
        strip.text.x = element_text(size=11, color = "black"),
        strip.placement = "outside", 
        strip.background = element_blank() )
```

```{r}
ggsave("~/Documents/air_polln_rna_seq/results/lm_MetWAS/mwas_pathway.tiff",
       dpi=300,
       bg="white",
       width = 9.5,
       height = 14)
```

# consolidated shapley style plot
```{r}
# Create your plot function (same as before)
create_pathway_plot <- function(data, pathway) {
  subset_data <- data %>% filter(super_pathway == pathway)
  
  ggplot(subset_data, aes(x = estimate, y = sub_pathway)) +
    geom_point(
      aes(color = case_when(
        fdr >= 0.05 ~ "lightgray",
        fdr < 0.05 & estimate > 0 ~ "blue",
        fdr < 0.05 & estimate < 0 ~ "red",
        TRUE ~ "lightgray"
      )), 
      size = 0.5, 
      position = position_jitter(height = 0.2, width = 0.1)
    ) +
    geom_vline(xintercept = 0, color = "black", linetype = "dotted") +
    scale_color_identity() +
    labs(x = NULL, y = NULL, title = pathway) +
    theme_minimal() +
    theme(
      plot.title = element_text(size = 8.3, color = "black", face = "bold"),
      axis.text.x = element_text(size = 8, colour = "black"),
      axis.text.y = element_text(size = 8, colour = "black"),
      panel.spacing = unit(1, "lines")
    )
}

# Open TIFF device
tiff(
  filename = "~/Documents/air_polln_rna_seq/results/lm_MetWAS/mwas_lm_summmary.tiff",
  width = 14,      # Width in inches
  height = 12,      # Height in inches
  units = "in",    # Units in inches
  res = 300,       # Resolution in dpi (300 for publication quality)
  compression = "lzw"  # Compression method
)

# Create your plots
lipid_plot <- create_pathway_plot(res_df, "Lipid")

col2_plots <- list(
  create_pathway_plot(res_df, "Amino Acid"),
  create_pathway_plot(res_df, "Carbohydrate"),
  create_pathway_plot(res_df, "Nucleotide")
)

col3_plots <- list(
  create_pathway_plot(res_df, "Cofactors and Vitamins"),
  create_pathway_plot(res_df, "Energy"),
  create_pathway_plot(res_df, "Partially Characterized Molecules"),
  create_pathway_plot(res_df, "Peptide"),
  create_pathway_plot(res_df, "Xenobiotics")
)

# Create column arrangements
col2 <- arrangeGrob(grobs = col2_plots, ncol = 1)
col3 <- arrangeGrob(grobs = col3_plots, ncol = 1)

# Create final arrangement
g <- arrangeGrob(
  lipid_plot, col2, col3,
  ncol = 3,
  widths = c(1.2, 1, 0.8),
  top = textGrob(" ", gp = gpar(fontsize = 14)),
  bottom = textGrob("Estimate", gp = gpar(fontsize = 12)),
  left = textGrob(" ", rot = 90, gp = gpar(fontsize = 14))
)

# Draw the plot
grid.draw(g)

# Close the device (this saves the file)
dev.off()
```

#Metabolie level plot
#load mwas results
```{r}
res_df<- read_csv("~/Documents/air_polln_rna_seq/results/lm_MetWAS/placenta_mwas_anno.csv")

#summary of results
summary_df <- res_df |>
  #filter(beta_dir %in% c("negative-significant", "positive-significant")) |>
  filter(fdr<0.05) |>
  mutate(prefix = ifelse(grepl("^pm_", variable), "pm_", 
                         ifelse(grepl("^bc_", variable), "bc_", NA))) |>
  filter(!is.na(prefix)) |>
  group_by(chemical_name, super_pathway, sub_pathway, prefix) |>
  summarise(count = n(), .groups = "drop") |>
  filter(count > 1)

viz_df<- res_df |>
  #filter(beta_dir %in% c("negative-significant", "positive-significant"))|>
  filter(fdr<0.05) |>
  mutate(exposure = case_when(
    str_starts(variable, "bc") ~ "Black carbon",
    str_starts(variable, "pm") ~ "PM2.5",
    TRUE ~ NA_character_  # Default case if neither condition is met
  ),
  exposure_window = gsub("^(pm_|bc_)", "", variable),
  exposure_window = factor(exposure_window, 
                             levels = sort(unique(as.numeric(gsub("^(pm_|bc_)", "", variable)))))
  ) |>
  filter(chemical_name %in% summary_df$chemical_name) |>
  mutate(
    predictor_numeric = as.numeric(gsub(".*_([0-9]+)$", "\\1", variable))
  ) |>
  mutate(super_pathway = fct_relevel(super_pathway, 
                                     "Lipid", 
                                     "Amino Acid", 
                                     "Carbohydrate", 
                                     "Cofactors and Vitamins", 
                                     "Peptide", 
                                     "Nucleotide", 
                                     "Xenobiotics"))
```

```{r}
ggplot(viz_df, aes(x = exposure_window, y = chemical_name, fill = estimate)) +
  geom_tile(color = "white") +
  scale_fill_gradient2(
  low = scales::muted("darkred"),
  mid = "gray90", 
  high = scales::muted("darkblue"),
  midpoint = 0,
  name = "Estimate"
) +
  labs(x = "Fine particulate matter exposure", y = "Metabolite") +
  facet_grid(super_pathway~exposure, scales="free", space="free",
             switch = "both",
             labeller = label_wrap_gen(width = 15)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, color="black"),
        axis.text.y = element_text(size=11, color="black"),
        legend.position = "bottom",
        legend.text = element_text(angle = 45, hjust = 1, color="black"),
        legend.title = element_text(vjust = 1),
        legend.key.height = unit(0.3, "cm"),
        panel.grid.major.x = element_blank(), 
        panel.grid.minor.x = element_blank() )
```

```{r}
ggsave("~/Documents/air_polln_rna_seq/results/lm_MetWAS/mwas_metabolite.tiff",
       dpi=300,
       bg="white",
       width = 10,
       height = 12)
```

#export metabolite list for yu-chin
```{r}
df_annotations_link <- df_annotations |>
  select(c(11,13:17)) |>
  mutate(pubchem = as.character(pubchem))

res_df<- read_csv("~/Documents/air_polln_rna_seq/results/lm_MetWAS/placenta_mwas_anno.csv")  |>
  filter(fdr < 0.20)|>
  mutate(
        air_pollutant = substr(variable, 1, 2),  # Extract first 2 characters
        predictor_end = ifelse(
            grepl("\\d+$", variable),  # Check if the string ends with numbers
            regmatches(variable, regexpr("\\d+$", variable)),  # Extract numeric part
            NA  # Set to NA if no numeric value at the end
        )
    ) |>
    group_by(air_pollutant, chemical_name, chem_id) |>
    summarise(
        count = n(),
        days_during_pregnancy = paste(na.omit(unique(predictor_end)), collapse = ","),  # Handle NA values gracefully
        .groups = "drop"
    ) |>
    arrange(air_pollutant, chemical_name, chem_id) |>
  left_join(df_annotations_link, by="chemical_name") |>
  select(c(1,4,5,2,3,6:10))

write_csv(res_df, "~/Documents/air_polln_rna_seq/results/for_yu_chin/placenta_MetWAS_anno_fdr_20pct.csv")
```



